draft-ietf-tls-external-psk-importer.md

```diff
@@ -99,53 +101,68 @@ a tuple of (Base Key, External Identity, KDF). The associated KDF (and hash func
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
+from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context {{CCB}}.
 
-ImportedIdentity.label MUST be bound to the protocol for which the key is imported. Thus,
-TLS 1.3 and QUICv1 {{!I-D.ietf-quic-transport}} MUST use "tls13" as the label. Similarly, TLS 1.2 and
-all prior TLS versions should use "tls12" as ImportedIdentity.label, as well as SHA256 as ImportedIdentity.hash.
+ImportedIdentity.protocol MUST be the (D)TLS protocol version for which the PSK is being imported.
+For example, TLS 1.3 and QUICv1 {{!I-D.ietf-quic-transport}} MUST use 0x0304, whereas TLS 1.2 uses 0x0303.
 Note that this means future versions of TLS will increase the number of PSKs derived from an external PSK.
 
-ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
-from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
-{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context [CCB].
+ImportedIdentity.kdf MUST be one of the KDF identifiers specified in {{kdf-registry}}.
 
 A unique and imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:
 
 ~~~
-   epskx = HKDF-Extract(0, epsk)
-   ipskx = HKDF-Expand-Label(epskx, "derived psk",
-                             Hash(ImportedIdentity), Hash.length)
+   epskx = KDF.Extract(0, epsk)
+   ipskx = KDF.Expand(epskx, concat("derived psk", Hash(ImportedIdentity)), KDF.Nh)
 ~~~
 
 [[TODO: The length of ipskx MUST match that of the corresponding and supported ciphersuites.]]
 
 The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK. It is not
-bound to ImportedIdentity.hash. If no hash function is specified, SHA-256 MUST be used.
-Differentiating epsk by ImportedIdentity.hash ensures that each imported PSK is only used with at most one
-hash function, thus satisfying the requirements in {{!RFC8446}}. Endpoints MUST import and derive an ipsk
-for each hash function used by each ciphersuite they support. For example, importing a key for
-TLS_AES_128_GCM_SHA256 and TLS_AES_256_GCM_SHA384 would yield two PSKs, one for SHA256 and another
-for SHA384. In contrast, if TLS_AES_128_GCM_SHA256 and TLS_CHACHA20_POLY1305_SHA256 are supported,
-only one derived key is necessary.
+bound to ImportedIdentity.kdf. If no hash function is specified, SHA-256 MUST be used.
```

Comment:
This is another example of where the KDF is confusing.  You should use something like "target KDF" and "EPSK KDF" to distinguish them.

Reply:
Yep, will do!